File: src/MyServiceBus/ScopeConsumerFactory.cs
using System;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;

namespace MyServiceBus;

public class ScopeConsumerFactory<TConsumer> : IConsumerFactory<TConsumer>
    where TConsumer : class
{
    readonly IServiceProvider provider;

    public ScopeConsumerFactory(IServiceProvider provider)
    {
        this.provider = provider;
    }

    [Throws(typeof(Exception))]
    public async Task Send<TMessage>(ConsumeContext<TMessage> context,
        IPipe<ConsumerConsumeContext<TConsumer, TMessage>> next) where TMessage : class
    {
        using var scope = provider.CreateScope();
        var contextProvider = scope.ServiceProvider.GetService<ConsumeContextProvider>();
        if (contextProvider != null)
            contextProvider.Context = context;

        try
        {
            var consumer = scope.ServiceProvider.GetRequiredService<TConsumer>();
            var consumerContext = new ConsumerConsumeContextImpl<TConsumer, TMessage>(consumer, context);
            await next.Send(consumerContext);
        }
        finally
        {
            if (contextProvider != null)
                contextProvider.Context = null;
        }
    }
}

File: src/Java/myservicebus/src/main/java/com/myservicebus/ScopeConsumerFactory.java
package com.myservicebus;

import java.util.concurrent.CompletableFuture;

import com.myservicebus.di.ServiceProvider;
import com.myservicebus.di.ServiceScope;

/**
 * Resolves consumers from a scoped service provider.
 */
public class ScopeConsumerFactory implements ConsumerFactory {
    private final ServiceProvider provider;

    public ScopeConsumerFactory(ServiceProvider provider) {
        this.provider = provider;
    }

    @Override
    public <TConsumer, T> CompletableFuture<Void> send(Class<TConsumer> consumerType,
            ConsumeContext<T> context,
            Pipe<ConsumerConsumeContext<TConsumer, T>> next) {
        try (ServiceScope scope = provider.createScope()) {
            ServiceProvider scoped = scope.getServiceProvider();
            ConsumeContextProvider ctxProvider = scoped.getService(ConsumeContextProvider.class);
            ctxProvider.setContext(context);
            try {
                @SuppressWarnings("unchecked")
                TConsumer consumer = (TConsumer) scoped.getService(consumerType);
                ConsumerConsumeContext<TConsumer, T> consumerContext = new ConsumerConsumeContext<>(consumer, context);
                return next.send(consumerContext);
            } finally {
                ctxProvider.clear();
            }
        }
    }
}

